### PR TITLE
use consistent argument parser

### PR DESCRIPTION
**Pull request by bekzod**
* aligns argument parsing logic together 
* fixes issue when `schedule/scheduleOnce/run/join` not behaving correctly when `[callback, string]` signature arguments passed

### CODE REVIEW

**Comment by bekzod** [comment]
Benchmark Results:

BEFORE: 
```
 Cancel - no timer .............................................................. 517,581,243.27 op/s
  Debounce & Cancel - function ..................................................... 1,289,046.93 op/s
  Debounce & Cancel - function, target ............................................. 1,301,786.83 op/s
  Later & Cancel - function .......................................................... 308,619.27 op/s
  Later & Cancel - function, target .................................................. 301,310.89 op/s
  Schedule & Cancel - function ..................................................... 3,900,332.93 op/s
  Schedule & Cancel - target, function ............................................. 1,752,119.60 op/s
  Schedule & Cancel - target, string method name ................................... 1,797,325.83 op/s
  Schedule & Cancel - target, string method name, 1 argument ....................... 1,699,961.99 op/s
  Schedule & Cancel - target, string method name, 2 arguments ...................... 1,625,160.63 op/s
  Schedule & Cancel - prescheduled, same queue - target, string method name .......... 978,925.97 op/s
  Schedule & Cancel - prescheduled, separate queue - target, string method name .... 1,698,020.15 op/s
  Throttle & Cancel - function .................................................... 11,009,497.78 op/s
  Throttle & Cancel - function, target ............................................ 10,748,428.83 op/s
```
AFTER:

```
  Cancel - no timer .............................................................. 523,623,083.22 op/s
  Debounce & Cancel - function ..................................................... 1,180,616.07 op/s
  Debounce & Cancel - function, target ............................................. 1,186,361.24 op/s
  Later & Cancel - function .......................................................... 303,471.09 op/s
  Later & Cancel - function, target .................................................. 297,463.47 op/s
  Schedule & Cancel - function ..................................................... 3,900,737.32 op/s
  Schedule & Cancel - target, function ............................................. 1,793,731.61 op/s
  Schedule & Cancel - target, string method name ................................... 1,825,686.05 op/s
  Schedule & Cancel - target, string method name, 1 argument ....................... 1,687,422.84 op/s
  Schedule & Cancel - target, string method name, 2 arguments ...................... 1,605,058.81 op/s
  Schedule & Cancel - prescheduled, same queue - target, string method name .......... 997,499.69 op/s
  Schedule & Cancel - prescheduled, separate queue - target, string method name .... 1,753,938.28 op/s
  Throttle & Cancel - function ..................................................... 8,365,828.54 op/s
  Throttle & Cancel - function, target ............................................. 7,986,201.60 op/s
```

File: lib/index.ts [hunk 1]
```diff
@@ -293,40 +352,9 @@ export default class Backburner {
...
-        target = args.shift();
-        method = target[args.shift()];
-      } else {
-        method = args.shift();
-      }
-    }
-
+  public later(...args);
+  public later() {
+    let [target, method, args, wait] = parseTimerArgs(...arguments);
```
Comment (stefanpenner) [inline lib/index.ts]: is the transpiled `loose` for this statement or does this try to do the funky iterables capable output?

File: lib/index.ts [hunk 2]
```diff
@@ -346,48 +374,15 @@ export default class Backburner {
...
-      if (isCoercableNumber(immediate)) {
-        wait = immediate;
-        isImmediate = true;
-      } else {
-        wait = args.pop();
-        isImmediate = immediate === true;
-      }
-    }
+  public throttle(): Timer {
+    let [target, method, args, wait, isImmediate = true] = parseDebounceArgs(...arguments);
```
Comment (stefanpenner) [inline lib/index.ts]: ^

File: lib/index.ts [hunk 3]
```diff
@@ -422,42 +417,8 @@ export default class Backburner {
...
-        isImmediate = false;
-      } else {
-        wait = args.pop();
-        isImmediate = immediate === true;
-      }
-    }
-
-    wait = parseInt(wait, 10);
+  public debounce(): Timer {
+    let [target, method, args, wait, isImmediate = false] = parseDebounceArgs(...arguments);
```
Comment (stefanpenner) [inline lib/index.ts]: ^

File: lib/index.ts (hunk 1 above)
Comment (stefanpenner) [reply lib/index.ts]: If its loose, we are good. If not, we should make it loose.

Comment (bekzod) [reply lib/index.ts]: its all good I think, here is output https://cl.ly/07260l1y0a2k/Screen%20Shot%202018-01-24%20at%2012.28.50%20PM.png

Comment (rwjblue) [reply lib/index.ts]: Ya, we should confirm how Ember consumes (it uses ES output so does its own transpile) also but I am pretty sure we use loose there too...

Comment (bekzod) [reply lib/index.ts]: yep same in ember https://cl.ly/3E0m1k3f463s/Screen%20Shot%202018-01-24%20at%202.19.44%20PM.png

**Comment by bekzod** [comment]
added tests

**Comment by bekzod** [comment]
things will look a lot cleaner :) after this 🎊🎊

**Pull request merged**